pkg/operators/cutter/cutter.go

```diff
@@ -0,0 +1,149 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cutter
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/datasource"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadget-service/api"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/operators"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/params"
+)
+
+const (
+	name         = "cutter"
+	ParamMaxRows = "max-rows"
+	Priority     = 9600
+)
+
+type cutterOperator struct{}
+
+func (s *cutterOperator) Name() string {
+	return name
+}
+
+func (s *cutterOperator) Init(params *params.Params) error {
+	return nil
+}
+
+func (s *cutterOperator) GlobalParams() api.Params {
+	return nil
+}
+
+func (s *cutterOperator) InstanceParams() api.Params {
+	return api.Params{
+		{
+			Key:   ParamMaxRows,
+			Title: "Max Rows",
+			Description: "The maximal number of rows to keep for each batch of data. " +
+				"If using multiple data sources, prefix the value with 'datasourcename:' and separate with ';'",
+		},
+	}
+}
+
+func (s *cutterOperator) InstantiateDataOperator(gadgetCtx operators.GadgetContext, instanceParamValues api.ParamValues) (operators.DataOperatorInstance, error) {
+	dsCutters := make(map[string]int)
+	for _, srt := range strings.Split(instanceParamValues[ParamMaxRows], ";") {
+		dsMaxRows := strings.Split(srt, ":")
+		dsName := ""
+		maxRowsStr := dsMaxRows[0]
+		if len(dsMaxRows) == 2 {
+			dsName = dsMaxRows[0]
+			maxRowsStr = dsMaxRows[1]
+		}
+		maxRows, err := strconv.Atoi(maxRowsStr)
+		if err != nil {
+			return nil, fmt.Errorf("parsing max-rows (%q): %w", maxRowsStr, err)
+		}
+		dsCutters[dsName] = maxRows
+	}
+	if len(dsCutters) == 0 {
+		return nil, nil
+	}
+
+	// Check edge cases
+	dsSpecific := true
+	if _, ok := dsCutters[""]; ok {
+		if len(dsCutters) > 1 {
+			return nil, fmt.Errorf("mixing max-rows rules with and without specifying data source")
+		}
+		dsSpecific = false
+	}
+
+	cutters := make(map[datasource.DataSource]int)
+	for _, ds := range gadgetCtx.GetDataSources() {
+		maxRows := dsCutters[ds.Name()]
+		if !dsSpecific {
+			maxRows = dsCutters[""]
+		}
+
+		if maxRows == 0 {
+			continue
+		}
+
+		if ds.Type() != datasource.TypeArray {
+			return nil, fmt.Errorf("max-rows can only be used on array data sources")
+		}
+
+		cutters[ds] = maxRows
+	}
+
+	return &cutterOperatorInstance{
+		cutters: cutters,
+	}, nil
+}
+
+func (s *cutterOperator) Priority() int {
+	return Priority
+}
+
+type cutterOperatorInstance struct {
+	cutters map[datasource.DataSource]int
+}
+
+func (s *cutterOperatorInstance) Name() string {
+	return name
+}
+
+func (s *cutterOperatorInstance) PreStart(gadgetCtx operators.GadgetContext) error {
+	for ds, maxRows := range s.cutters {
+		ds.SubscribeArray(func(ds datasource.DataSource, data datasource.DataArray) error {
+			if data.Len() <= maxRows {
+				return nil
+			}
+
+			// TODO: Cut the data
```

Comment:
TODO: Should `DataArray` implement a `Drop` method?

Reply:
Maybe rather a `.Resize()` (as that's more generic) that can grow and shrink later on - for now, we only need to implement shrinking and returning an error if trying to grow. It needs to release the removed elements, then.

Reply:
Makes sense as idea. Let me give it a try.